### PR TITLE
LOK-2643 : BE     Expose Graphql service to get data for cpu and memory utilization

### PR DESCRIPTION
**Pull request by aurang-zeb313**
## Description
BE: Expose Graphql service to get data for CPU and Memory utilization.

## Jira link(s)
- https://opennms.atlassian.net/browse/LOK-2643

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

**Comment by aurang-zeb313** [comment]
@cgorantla  as discussed the same expression is working using curl but not working through graphql UI 
`curl --location --globoff 'http://localhost:9000/prometheus/api/v1/query_range' \
--header 'Accept: application/json' \
--header 'X-Scope-OrgID: opennms-prime' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'query=100 - (CpuRawIdle{monitor="SNMP",node_id="1"} / (CpuRawIdle{monitor="SNMP",node_id="1"} +CpuRawInterrupt{monitor="SNMP",node_id="1"} +CpuRawUser{monitor="SNMP",node_id="1"} +CpuRawWait{monitor="SNMP",node_id="1"} +CpuRawNice{monitor="SNMP",node_id="1"} +CpuRawSystem{monitor="SNMP",node_id="1"} +CpuRawKernel{monitor="SNMP",node_id="1"} +CpuRawSoftIRQ{monitor="SNMP",node_id="1"}  +CpuRawSteal{monitor="SNMP",node_id="1"} +CpuRawGuest{monitor="SNMP",node_id="1"} +CpuRawGuestNice{monitor="SNMP",node_id="1"})) * 100' \
--data-urlencode 'start=2024-07-03T12:30:55Z' \
--data-urlencode 'end=2024-07-04T12:32:13Z' \
--data-urlencode 'step=10' \
--data-urlencode 'timeout=0'`

**Review by cgorantla — COMMENTED** [review]
Use grafana to validate the queries.  Check whether generated query works with grafana.  I didn't get any result back on grafana with the generated query from GraphQL although individual metrics in that query seems to work.

**Comment by aurang-zeb313** [comment]
> Use grafana to validate the queries. Check whether generated query works with grafana. I didn't get any result back on grafana with the generated query from GraphQL for cpu utilization although individual metrics in that query seems to work.

Yes i have validated it using grafana and it give me the result. As you can see in snap.

Screenshot from 2024-07-11 17-16-54

**Comment by aurang-zeb313** [comment]
> Use grafana to validate the queries. Check whether generated query works with grafana. I didn't get any result back on grafana with the generated query from GraphQL for cpu utilization although individual metrics in that query seems to work.

As discussed the issue is resolved by encoding the request body. When we have a large request body we need to encode it for the range-query

File: rest-server/src/main/java/org/opennms/horizon/server/service/metrics/Constants.java [hunk 1]
```diff
@@ -79,4 +81,17 @@ public final class Constants {
                     unless
                 count(irate(ifHCOutOctets[4m])) == 0 and count(sum_over_time(network_out_total_bytes[1m])) == 0
         """;
+
+    public static final String QUERY_FOR_CPU_UTILIZATION =
```
Comment (synqotik) [inline rest-server/src/main/java/org/opennms/horizon/server/service/metrics/Constants.java:85]: I'm still a bit unclear on these expressions. I would consider reformatting them a bit more and add more parentheses to make it more clear what is the numerator and what is the denominator, and not rely on operator precedence.

**Review by synqotik — DISMISSED** [review]
Front End part looks good. We will probably eventually use constants for the metric names, but we can do that later once we have more metrics.

Java part seems good, but I think the expressions can be made more clear, I'm not quite sure what the numerator and denominator are intended to be.

File: rest-server/src/main/java/org/opennms/horizon/server/service/metrics/Constants.java [hunk 2]
```diff
@@ -79,4 +81,17 @@ public final class Constants {
...
+            """
+         ( 1 - avg_over_time(CpuRawIdle%1$s[24h]) / (avg_over_time(CpuRawIdle%1$s[24h]) + avg_over_time(CpuRawInterrupt%1$s[24h]) + avg_over_time(CpuRawUser%1$s[24h]) + avg_over_time(CpuRawWait%1$s[24h]) + avg_over_time(CpuRawNice%1$s[24h]) + avg_over_time(CpuRawSystem%1$s[24h]) + avg_over_time(CpuRawKernel%1$s[24h]) + avg_over_time(CpuRawSoftIRQ%1$s[24h])  + avg_over_time(CpuRawSteal%1$s[24h]) + avg_over_time(CpuRawGuest%1$s[24h]) + avg_over_time(CpuRawGuestNice%1$s[24h])))  * 100
```
Comment (cgorantla) [inline rest-server/src/main/java/org/opennms/horizon/server/service/metrics/Constants.java:87]: avg_over_time over 24h is not the way to calculate instantaneous CPU utilization.

File: rest-server/src/main/java/org/opennms/horizon/server/service/metrics/Constants.java [hunk 3]
```diff
@@ -79,4 +81,17 @@ public final class Constants {
...
+        """;
+
+    public static final String QUERY_FOR_MEMORY_UTILIZATION =
+            """
+               (1 -  ((avg_over_time(memTotalReal%1$s[24h]) - avg_over_time(memAvailReal%1$s[24h]))
```
Comment (cgorantla) [inline rest-server/src/main/java/org/opennms/horizon/server/service/metrics/Constants.java:92]: avg_over_time over 24h is not the way to calculate instantaneous memory utilization.

File: rest-server/src/main/java/org/opennms/horizon/server/service/metrics/Constants.java [hunk 4]
```diff
@@ -79,4 +81,17 @@ public final class Constants {
                     unless
                 count(irate(ifHCOutOctets[4m])) == 0 and count(sum_over_time(network_out_total_bytes[1m])) == 0
         """;
+
+    public static final String QUERY_FOR_CPU_UTILIZATION =
+            """
+       (100 -  (CpuRawIdle%1$s / (CpuRawIdle%1$s + CpuRawInterrupt%1$s + CpuRawUser%1$s + CpuRawWait%1$s + CpuRawNice%1$s + CpuRawSystem%1$s + CpuRawKernel%1$s + CpuRawSoftIRQ%1$s  + CpuRawSteal%1$s + CpuRawGuest%1$s + CpuRawGuestNice%1$s))  * 100 )
```
Comment (cgorantla) [inline rest-server/src/main/java/org/opennms/horizon/server/service/metrics/Constants.java:87]: Looks like these are counters, so we have to take the difference between two successive values.

File: rest-server/src/main/java/org/opennms/horizon/server/service/metrics/QueryService.java [hunk 5]
```diff
@@ -238,13 +245,27 @@ public String getLabelsQueryString(Map<String, String> labels) {
         return sb.toString();
     }
 
+    private void appendLabels(StringBuilder sb, String monitorValue, String nodeIdValue) {
```
Comment (cgorantla) [inline rest-server/src/main/java/org/opennms/horizon/server/service/metrics/QueryService.java:248]: looks like it's not used anywhere.

File: rest-server/src/main/java/org/opennms/horizon/server/service/metrics/QueryService.java [hunk 6]
```diff
@@ -238,13 +245,27 @@ public String getLabelsQueryString(Map<String, String> labels) {
...
+    }
+
     private String addTimeRange(Integer timeRange, TimeRangeUnit timeRangeUnit, String queryString) {
         if (timeRange != null && timeRangeUnit != null) {
             return queryString + "[" + timeRange + timeRangeUnit.value + "]";
         }
         return queryString;
     }
 
+    private String addMetricTimeRange(Integer timeRange, TimeRangeUnit timeRangeUnit, String queryString) {
```
Comment (cgorantla) [inline rest-server/src/main/java/org/opennms/horizon/server/service/metrics/QueryService.java:262]: looks like it's not used anywhere.

**Review by cgorantla — APPROVED** [review]
haven't validated but LGTM.

File: rest-server/src/main/java/org/opennms/horizon/server/service/metrics/QueryService.java (hunk 6 above)
Comment (aurang-zeb313) [reply rest-server/src/main/java/org/opennms/horizon/server/service/metrics/QueryService.java]: Yes, addMetricTimeRange removed

File: rest-server/src/main/java/org/opennms/horizon/server/service/metrics/QueryService.java (hunk 5 above)
Comment (aurang-zeb313) [reply rest-server/src/main/java/org/opennms/horizon/server/service/metrics/QueryService.java]: Yes, removed

**Pull request merged**